Add Challenge 16: Iterator (#260)
model-checking/verify-rust-std: doc/src/challenges/0016-iter.md

> # Challenge 16: Verify the safety of Iterator functions

Reviewer: This PR is missing an update to `doc/src/SUMMARY.md` to link to this new challenge.

Author: I will create a separate PR for that file with all the added challenges.

Reviewer 2: Why? I don't think we should merge incomplete PRs; let's just have the changes we need in a single one.

- **Status:** Open

> - **Tracking Issue:** [#29](https://github.com/model-checking/verify-rust-std/issues/29)

Reviewer: Is that really the right URL? I would have thought you'd want to create a fresh tracking issue. When you do so, can you, as discussed offline, please include a checklist with all the functions listed below?

Author: Yes, it is just a old URL, I will create the tracking issue and fix the prize one after the challenge is approved.

- **Start date:** *2025/03/07*

> - **End date:** *2025/10/17*

Reviewer: Please use ISO data formatting (2025-03-07, 2025-10-17).

> - **Reward:** *5000 USD*

Reviewer 2: Note: we need to confirm this reward amount before merging.

Reviewer 2: Per offline discussion, it seems like now that we have two technical approvals for this challenge, we need the reward committee to approve this amount. So I am unresolving this conversation; we should not merge this challenge until we get the go-ahead.

-------------------

## Goal

Verify the safety of iter functions that are defined in (library/core/src/iter/adapters):

### Success Criteria

> Write and prove the contract for the safety of the following functions:

Reviewer 2: Some of these functions are safe; we shouldn't ask for safety contracts for safe functions. If they're safe abstractions, we can still ask them to verify that there's no UB, but please separate these into unsafe functions (which should have contracts to specify their preconditions) and safe functions (which should not need contracts for safety).

| Function | Defined in |
|---------| ---------|
|next_back_remainder| array_chunks.rs|
|fold| array_chunks.rs|

> |__iterator_get_unchecked| clone.rs|

Reviewer 2: Hm, I think it may make more sense to just ask for the callers of `__iterator_get_unchecked`--the double `__` makes me think this is meant to be super internal to `core` and not something that would benefit much from an upstream contract.
Open to disagreement, though.

Reviewer 2: Okay, since we're doing this in the other challenges I actually think it's fine to leave it.

|fold| clone.rs|
|next_unchecked| clone.rs|
|__iterator_get_unchecked| copied.rs|
|spec_next_chunk| copied.rs|
|__iterator_get_unchecked| enumerate.rs|
|next_chunk| filter.rs|
|next_chunk| filter_map.rs|
|__iterator_get_unchecked | fuse.rs|
|__iterator_get_unchecked | map.rs|
|next_unchecked | map.rs|
|as_array_ref | map_windows.rs|
|as_uninit_array_mut | map_windows.rs|
|push | map_windows.rs|
|drop | map_windows.rs|
|__iterator_get_unchecked | skip.rs|
|original_step | step_by.rs|
|spec_fold| take.rs|
|spec_for_each| take.rs|
|__iterator_get_unchecked | zip.rs|
|get_unchecked| zip.rs|
|fold| zip.rs|
|next| zip.rs|
|nth| zip.rs|
|next_back| zip.rs|
|spec_fold| zip.rs|

The verification must be unbounded---it must hold for slices of arbitrary length.

> The verification must be hold for generic type `T` (no monomorphization).

Reviewer: ```suggestion
The verification must hold for generic type `T` (no monomorphization).
```

# Challenge 16: Verify the safety of Iterator functions

- **Status:** Open
- **Tracking Issue:** [#280](https://github.com/model-checking/verify-rust-std/issues/280)
- **Start date:** *2025-03-07*
- **End date:** *2025-10-17*

> - **Reward:** *5000 USD*

Reviewer 2: ```suggestion
- **Reward:** *10,000 USD*
```

### Success Criteria

Write and prove the contract for the safety of the following unsafe functions:

| Function | Defined in |
|---------| ---------|
|__iterator_get_unchecked| clone.rs|
|next_unchecked| clone.rs|
|__iterator_get_unchecked| copied.rs|
|__iterator_get_unchecked| enumerate.rs|
|__iterator_get_unchecked | fuse.rs|
|__iterator_get_unchecked | map.rs|
|next_unchecked | map.rs|
|__iterator_get_unchecked | skip.rs|
|__iterator_get_unchecked | zip.rs|
|get_unchecked| zip.rs|

> Prove the absence of UB for following safe functions:

Reviewer 2: ```suggestion
Prove the absence of undefined behavior for following safe abstractions:
```
